[Darwin] Further restrict inference of the simulator environment

DougGregor opened the pull request:
The *-simulator target triples have been used mostly consistently in build systems for
several years to indicate simulator targets... except for Swift's own CMake configuration
and `build-script`, which was still not doing it until recently. There are still `.swiftinterface`
files around that depend on the inference.

Rather than outright removing inference of the "simulator" environment, move it into
the driver and warn that it's happening, so other tools can move off it over time. This
ensures that all frontend/linker/etc. jobs see the "simulator" environment, so the hack is
only in a single place.

This is as far as we can go toward completely removing the hack (rdar://problem/35810403),
because of the need to continue supporting existing `.swiftinterface` files.

DougGregor committed:
[Platform] Stop inferring simulator-ness.

The *-simulator target triples have been used consistently in tools for
several years to indicate simulator targets. Stop inferring the
simulator part, rdar://problem/35810403.

DougGregor commented:
@swift-ci please test

harlanhaskins commented:
Seems like we should also remove this from `swift-driver`. https://github.com/apple/swift-driver/blob/master/Sources/SwiftDriver/Utilities/Triple%2BPlatforms.swift#L133

DougGregor commented:
Linux failure is unrelated. Trying again...

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please smoke test Linux

DougGregor commented:
GitHub connectivity problems affected the compatibility suite. Sigh.

DougGregor commented:
Sigh. I need to split this pull request into "fix our build system / test" and "remove the cruft".

DougGregor commented:
@swift-ci please test

DougGregor committed:
[Serialization] Require module files to match on environment, too.

This prevents mixing simulator and non-simulator modules.

DougGregor commented:
@swift-ci please clean test

DougGregor commented:
@swift-ci please clean test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please clean test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor commented:
@swift-ci please test

DougGregor commented:
Much of this pull request came in via https://github.com/apple/swift/pull/30843. I'll rebase the rest on top of this.

compnerd commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -172,6 +174,13 @@ macro(configure_sdk_darwin
     SWIFT_SDK_${prefix}_MODULE_ARCHITECTURES)     # result
 
+  # Determine whether this is a simulator SDK.
+  if ( ${xcrun_name} MATCHES "simulator" )
```
The convention in LLVM (and Swift) is not have spaces after the `if` and not to have spaces inside the `(`.
```suggestion
  if(xcrun_name MATCHES simulator)
```

DougGregor replied:
I had no idea, thank you!

compnerd commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -181,6 +190,12 @@ macro(configure_sdk_darwin
         "${arch}-apple-${SWIFT_SDK_${prefix}_TRIPLE_NAME}")
 
+    # If this is a simulator target, append -simulator.
+    if (SWIFT_SDK_${prefix}_IS_SIMULATOR)
```
```suggestion
    if(SWIFT_SDK_${prefix}_IS_SIMULATOR)
```

compnerd commented on cmake/modules/SwiftConfigureSDK.cmake:
```diff
@@ -181,6 +190,12 @@ macro(configure_sdk_darwin
+    if (SWIFT_SDK_${prefix}_IS_SIMULATOR)
+      set(SWIFT_SDK_${prefix}_ARCH_${arch}_TRIPLE
+          "${SWIFT_SDK_${prefix}_ARCH_${arch}_TRIPLE}-simulator")
+    endif ()
```
```suggestion
    endif()
```

DougGregor committed:
[Basic] Remove swift::tripleIsAnySimulator().

It's just Triple::isSimulatorEnvironment() now.

DougGregor committed:
[Driver] Infer simulator-ness for iOS/tvOS/watchOS.

Localize the hack to infer simulator-ness of the target in the driver
itself, when it first processes the target. Emit a warning about the
missing "-simulator" and correct the triple immediately.

This gives a longer grace period for tools that might still not pass
through the simulator environment, while narrowing the hack.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Driver] Update a test for the simulator inference.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Frontend] Infer simulator environment in the frontend, too.

Some code paths that see target triples go through the frontend
without seeing the driver. Therefore, perform the same "simulator"
inference for x86 iOS/tvOS/watchOS triples also in the frontend,
to ensure that we remain compatible. Also make sure that
-print-target-info performs the appropriate adjustment.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test source compatibility

DougGregor committed:
[Simulators] Only infer simulator environment when there is no environment.

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci please test

DougGregor commented:
@swift-ci test platform Linux

DougGregor commented:
@swift-ci smoke test Linux

DougGregor commented:
@swift-ci smoke test Linux

fredriss commented:
When this was merged, the Linux PR testing started failing in lldb tests. First Failure here: https://ci.swift.org/job/oss-swift-package-linux-ubuntu-16_04/4933/

I have no idea what's going on, it doesn't make much sense, but the revert unblocked things.
